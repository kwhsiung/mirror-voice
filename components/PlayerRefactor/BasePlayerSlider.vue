<template>
  <VueSlider
    v-model="value"
    class="slider"
    :style="{ padding: '0' }"
    :rail-style="{ background }"
    v-bind="options"
  />
</template>

<script>
import VueSlider from 'vue-slider-component/dist-css/vue-slider-component.umd.min.js'

export default {
  components: {
    VueSlider
  },
  props: {
    direction: {
      type: String,
      required: true,
      validator(value) {
        return ['horizontal', 'vertical'].includes(value)
      }
    },
    valueBuffered: {
      type: Number,
      default: 0,
      require: true,
      validator(value) {
        return value >= 0 && value <= 1
      }
    },
    valueCurrent: {
      type: Number,
      default: 0,
      require: true,
      validator(value) {
        return value >= 0 && value <= 1
      }
    }
  },
  computed: {
    value: {
      get() {
        return this.valueCurrent
      },
      set(value) {
        this.$emit('update:valueCurrent', value)
      }
    },
    options() {
      return {
        interval: 0.001,
        direction: this.direction === 'horizontal' ? 'ltr' : 'btt',
        width: this.direction === 'horizontal' ? 'auto' : 4,
        height: this.direction === 'horizontal' ? 4 : 126,
        min: 0,
        max: 1,
        tooltip: 'none',
        dotSize: 12,
        processStyle: {
          backgroundColor: '#d84939'
        }
      }
    },
    background() {
      const buffered = this.valueBuffered * 100
      return `linear-gradient(90deg, #7d7d7d ${buffered}%, black ${buffered}%)`
    }
  }
}
</script>

<style lang="stylus" scoped>
.slider
  cursor pointer
  position relative
  & >>> .vue-slider-dot
    background-color #eeeeee
    border-radius 100%
    position absolute
  & >>> .vue-slider-rail
    height 100%
</style>
