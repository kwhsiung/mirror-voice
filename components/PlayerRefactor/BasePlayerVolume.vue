<template>
  <div
    v-click-outside="hideSlider"
    class="volume"
  >
    <button
      :class="[
        'button-toggler',
        { 'mute': shouldMute }
      ]"
    >
      <img
        v-show="shouldMute"
        class="button-toggler__img"
        src="~/assets/img/btn_volume_min.png"
        alt="volume min"
        @click="toggleSlider"
      >
      <img
        v-show="!shouldMute"
        class="button-toggler__img"
        src="~/assets/img/btn_volume_normal.png"
        alt="volume normal"
        @click="toggleSlider"
      >
    </button>
    <div
      v-show="showSlider"
      class="volume__slider slider"
    >
      <BasePlayerSlider
        :direction="'vertical'"
        :value-current.sync="volume"
      />
    </div>
  </div>
</template>

<script>
import BasePlayerSlider from './BasePlayerSlider.vue'

export default {
  components: {
    BasePlayerSlider
  },
  props: {
    volumeCurrent: {
      type: Number,
      default: 1,
      validator(value) {
        return value >= 0 && value <= 1
      }
    }
  },
  data() {
    return {
      showSlider: false
    }
  },
  computed: {
    volume: {
      get() {
        return this.volumeCurrent
      },
      set(value) {
        this.$emit('update:volumeCurrent', value)
      }
    },
    shouldMute() {
      return this.volume <= 0
    }
  },
  methods: {
    toggleSlider() {
      this.showSlider = !this.showSlider
    },
    hideSlider() {
      this.showSlider = false
    }
  }
}
</script>

<style lang="stylus" scoped>
.volume
  height 100%
  display flex
  justify-content center
  align-items center
  position relative
  &__slider
    position absolute
    bottom calc(100% + 3px)
    width 52px
    height 164px
    background-color #313131
    border-radius 2px

.button-toggler
  height 18px
  cursor pointer
  padding 0
  background-color transparent
  border none
  outline none
  &__img
    height 100%

.slider
  display flex
  justify-content center
  align-items center
</style>
