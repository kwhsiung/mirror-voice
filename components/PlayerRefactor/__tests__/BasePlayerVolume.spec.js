import { shallowMount } from '@vue/test-utils'
import BasePlayerVolume from '../BasePlayerVolume.vue'
import BasePlayerSlider from '../BasePlayerSlider.vue'

describe('BasePlayerVolume', () => {
  test('have "mute" class on .button-toggler if volumeCurrent props is 0, otherwise not', () => {
    let wrapper = shallowMount(BasePlayerVolume, {
      propsData: {
        volumeCurrent: 0
      }
    })
    let img = wrapper.find('.button-toggler')

    expect(img.classes()).toContain('mute')

    wrapper = shallowMount(BasePlayerVolume, {
      propsData: {
        volumeCurrent: 0.66
      }
    })
    img = wrapper.find('.button-toggler')

    expect(img.classes()).not.toContain('mute')
  })

  test('emit "update:volumeCurrent" event while slider emit "update:valueCurrent" with value', () => {
    const valueChanged = 0.88
    const wrapper = shallowMount(BasePlayerVolume)
    const slider = wrapper.find(BasePlayerSlider)
    slider.vm.$emit('update:valueCurrent', valueChanged)

    expect(wrapper.emitted('update:volumeCurrent')[0]).toEqual([valueChanged])
  })
})

describe('snapshot tests', () => {
  test('renders component correctly if we provide efficient props', () => {
    const wrapper = shallowMount(BasePlayerVolume, {
      propsData: {
        volumeCurrent: 0.66
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })

  test('renders component correctly if we provide efficient props and is zero', () => {
    const wrapper = shallowMount(BasePlayerVolume, {
      propsData: {
        volumeCurrent: 0
      }
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
