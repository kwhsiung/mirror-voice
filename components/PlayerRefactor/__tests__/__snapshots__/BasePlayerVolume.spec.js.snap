// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot tests renders component correctly if we provide efficient props 1`] = `
<div
  class="volume"
>
  <button
    class="button-toggler"
  >
    <img
      alt="volume min"
      class="button-toggler__img"
      src="~/assets/img/btn_volume_min.png"
      style="display: none;"
    />
     
    <img
      alt="volume normal"
      class="button-toggler__img"
      src="~/assets/img/btn_volume_normal.png"
    />
  </button>
   
  <div
    class="volume__slider slider"
    style="display: none;"
  >
    <baseplayerslider-stub
      direction="vertical"
      valuebuffered="0"
      valuecurrent="0.66"
    />
  </div>
</div>
`;

exports[`snapshot tests renders component correctly if we provide efficient props and is zero 1`] = `
<div
  class="volume"
>
  <button
    class="button-toggler mute"
  >
    <img
      alt="volume min"
      class="button-toggler__img"
      src="~/assets/img/btn_volume_min.png"
    />
     
    <img
      alt="volume normal"
      class="button-toggler__img"
      src="~/assets/img/btn_volume_normal.png"
      style="display: none;"
    />
  </button>
   
  <div
    class="volume__slider slider"
    style="display: none;"
  >
    <baseplayerslider-stub
      direction="vertical"
      valuebuffered="0"
      valuecurrent="0"
    />
  </div>
</div>
`;
